hwilib/devices/onekey.py

```diff
@@ -0,0 +1,362 @@
+        self.auto_lock_delay_ms = auto_lock_delay_ms
+        self.display_rotation = display_rotation
+        self.experimental_features = experimental_features
+        self.offset = offset
+        self.ble_name = ble_name
+        self.ble_ver = ble_ver
+        self.ble_enable = ble_enable
+        self.se_enable = se_enable
+        self.se_ver = se_ver
+        self.backup_only = backup_only
+        self.onekey_version = onekey_version
+        self.onekey_serial = onekey_serial
+        self.bootloader_version = bootloader_version
+        self.serial_no = serial_no
+        self.boardloader_version = boardloader_version
+
+
+DEFAULT_MAPPING.register(OnekeyFeatures)
+
+USB_IDS = {(0x1209, 0x4F4A), (0x1209, 0x4F4B), }
+
+ONEKEY_LEGACY = TrezorModel(
+    name="1",
+    minimum_version=(2, 11, 0),
+    vendors=VENDORS,
+    usb_ids=USB_IDS,
+    default_mapping=DEFAULT_MAPPING,
+)
+
+ONEKEY_TOUCH = TrezorModel(
+    name="T",
+    minimum_version=(4, 2, 0),
+    vendors=VENDORS,
+    usb_ids=USB_IDS,
+    default_mapping=DEFAULT_MAPPING,
+)
+
+ONEKEYS = (ONEKEY_LEGACY, ONEKEY_TOUCH)
+
+
+def model_by_name(name: str) -> Optional[TrezorModel]:
+    for model in ONEKEYS:
+        if model.name == name:
+            return model
+    return None
+
+
+# ===============overwrite methods for onekey device begin============
+
+
+def _refresh_features(self: object, features: Features) -> None:
+    """Update internal fields based on passed-in Features message."""
+    if not self.model:
+        self.model = model_by_name(features.model or "1")
+        if self.model is None:
+            raise RuntimeError("Unsupported OneKey model")
+
+    if features.vendor not in self.model.vendors:
+        raise RuntimeError("Unsupported device")
+    self.features = features
+    self.version = (*map(int, self.features.onekey_version.split(".")), )
+    self.check_firmware_version(warn_only=True)
+    if self.features.session_id is not None:
+        self.session_id = self.features.session_id
+        self.features.session_id = None
+
+def button_request(self: object, code: Optional[int]) -> None:
+    if not self.prompt_shown:
+        print("Please confirm action on your OneKey device", file=sys.stderr)
+    if not self.always_prompt:
+        self.prompt_shown = True
+
+
+# ===============overwrite methods for onekey device end============
+
+ONEKEY_EMULATOR_PATH = "127.0.0.1:54935"
+class OnekeyClient(TrezorClient):
+    def __init__(
+        self,
+        path: str,
+        password: Optional[str] = None,
+        expert: bool = False,
+        chain: Chain = Chain.MAIN,
+    ) -> None:
+        super().__init__(path, password, expert, chain, webusb_ids=USB_IDS, sim_path=ONEKEY_EMULATOR_PATH)
+        self.client._refresh_features = MethodType(_refresh_features, self.client)
+        if not isinstance(self.client.ui, debuglink.DebugUI):
+            self.client.ui.button_request = MethodType(button_request, self.client.ui)
```

Comment:
What is the reason for overriding `button_request`? It seems the only difference is the string, in which case the original string can instead be generalized rather than duplicating code.